exec/java-exec/src/main/java/org/apache/drill/exec/expr/EqualityVisitor.java

```diff
@@ -0,0 +1,322 @@
+    if (!(value instanceof Decimal38Expression)) {
+      return false;
+    }
+    if (decExpr.getBigDecimal() != ((Decimal38Expression) value).getBigDecimal()) {
+      return false;
+    }
+    if (!decExpr.getMajorType().equals(((Decimal38Expression) value).getMajorType())) {
+      return false;
+    }
+    return false;
+  }
+
+  @Override
+  public Boolean visitDoubleConstant(DoubleExpression dExpr, LogicalExpression value) throws RuntimeException {
+    if (!(value instanceof DoubleExpression)) {
+      return false;
+    }
+    return dExpr.getDouble() == ((DoubleExpression) value).getDouble();
+  }
+
+  @Override
+  public Boolean visitBooleanConstant(BooleanExpression e, LogicalExpression value) throws RuntimeException {
+    if (!(value instanceof BooleanExpression)) {
+      return false;
+    }
+    return e.getBoolean() == ((BooleanExpression) value).getBoolean();
+  }
+
+  @Override
+  public Boolean visitQuotedStringConstant(QuotedString e, LogicalExpression value) throws RuntimeException {
+    if (!(value instanceof QuotedString)) {
+      return false;
+    }
+    return e.getString().equals(((QuotedString) value).getString());
+  }
+
+  @Override
+  public Boolean visitNullExpression(NullExpression e, LogicalExpression value) throws RuntimeException {
+    if (!(value instanceof NullExpression)) {
+      return false;
+    }
+    return e.getMajorType().equals(value.getMajorType());
+  }
+
+  @Override
+  public Boolean visitCastExpression(CastExpression e, LogicalExpression value) throws RuntimeException {
+    if (!(value instanceof CastExpression)) {
+      return false;
+    }
+    if (!e.getMajorType().equals(value.getMajorType())) {
+      return false;
+    }
+    return checkChildren(e, value);
+  }
+
+  @Override
+  public Boolean visitConvertExpression(ConvertExpression e, LogicalExpression value) throws RuntimeException {
+    if (!(value instanceof ConvertExpression)) {
+      return false;
+    }
+    if (!e.getConvertFunction().equals(((ConvertExpression) value).getConvertFunction())) {
+      return false;
+    }
+    return checkChildren(e, value);
+  }
+
+  @Override
+  public Boolean visitNullConstant(TypedNullConstant e, LogicalExpression value) throws RuntimeException {
+    if (!(value instanceof TypedNullConstant)) {
+      return false;
+    }
+    return e.getMajorType().equals(e.getMajorType());
+  }
+
+  @Override
+  public Boolean visitUnknown(LogicalExpression e, LogicalExpression value) throws RuntimeException {
+    if (e instanceof ValueVectorReadExpression && value instanceof ValueVectorReadExpression) {
+      return visitValueVectorReadExpression((ValueVectorReadExpression) e, (ValueVectorReadExpression) value);
+    }
+    return false;
+  }
+
+  private Boolean visitValueVectorReadExpression(ValueVectorReadExpression e, ValueVectorReadExpression value) {
+    return e.getTypedFieldId().equals(value.getTypedFieldId());
```

Comment:
Please post the revised patch when it's ready.

In the meantime, I feel it's probably not the best practice that EqualityVisitor relies on the cleaning of cache whenever MappingSet is changed.   For one thing, people would expect EqualityVisitor return true if and only if two expressions are identical by semantics; it should not return true for two semantically different expressions, and only make it work by assuming the cache is cleaned.

Also, if people use Equality for ExpressionHolder later on for other use case, and did not know such trick, then it's likely some bug would be encountered.